overlord/snapstate/handlers.go

```diff
@@ -1150,9 +1150,43 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		return copyDataErr
 	}
 
-	// the migration hasn't been done - do it now
-	if opts.UseHidden && !opts.MigratedToHidden {
-		if err = m.backend.HideSnapData(snapsup.InstanceName()); err != nil {
+	var oldBase string
+	if oldInfo != nil {
+		oldBase = oldInfo.Base
+	}
+
+	// snap is already migrated, the flag has been turned off but we're
+	// refreshing to a core22 based revision for the 1st time
```

Comment:
this is mixing predicates helpers and predicates in the ifs,  maybe it would be better to have a single function that returns a couple of booleans? or return an enumeration of possible states/actions and then there's  a switch here?

just a suggestion but this is a bit hard to follow